### ISSUE TITLE
Incorrect SubtrajectoryReplayBuffer behavior on truncation/termination?

### ISSUE THREAD

**Issue by adrian-schucht** [mlaux1/rl-blox#213]
See #212 

---

In `SubtrajectoryReplayBuffer` (`rl_blox/blox/replay_buffer.py`), it seems that the `mask_`, which masks out potential starting indices for subtrajectories so they will not be used for samples, is changed incorrectly: When a truncation transition is added, the mask remains unchanged, as the indices masked out (set to 0) were masked out before already. When a termination transition is added, the code masks in starting indices that should not be masked in, which leads to `sample_batch()` producing subtrajectories that do not respect the border between episodes, i.e., a subtrajectory may start in one episode and end in the next, including transitions from both.

```python
...
self.mask_[self.insert_idx % self.buffer_size] = 0
past_idx = (
    self.insert_idx
    - np.arange(min(self.episode_timesteps, self.horizon))
    - 1
) % self.buffer_size
self.mask_[past_idx] = (
    0 if sample["truncated"] else 1
)  # mask out truncated subtrajectories
...
```
See the lines in context [here](https://github.com/adrian-schucht/rl-blox/blob/b856c20081cb27f263c99951b1a38f213e3bc1ff/rl_blox/blox/replay_buffer.py#L255-L263).

_The actual issue:_ Tests were added that confirm this behavior. Most importantly, **see the last test, which shows that trajectories may contain more than one episode's transitions**: `test_subtrajectory_replay_buffer_respects_episode_boundaries_on_termination(...)`. This test works by adding samples from one episode with non-negative observations, then terminating this episode and continuing to add samples from the next episode with *negative* observations, before sampling from the buffer to show that the sampled trajectories contain both non-negative and negative observations. Relevant test output:

```
______________________________ test_subtrajectory_replay_buffer_termination ______________________________

strb_for_masking = <rl_blox.blox.replay_buffer.SubtrajectoryReplayBuffer object at 0x7cb036bbc230>

    def test_subtrajectory_replay_buffer_termination(strb_for_masking):
        buf = strb_for_masking
        add(buf, obs=4, next_obs=5, term=True, trunc=False)
        assert (buf.current_len == 6)
>       assert (buf.mask_ == jnp.array([1, 1, 0, 0, 0, 0, 0, 0, 0, 0])).all()
E       assert Array(False, dtype=bool)
E        +  where Array(False, dtype=bool) = _all()
E        +    where _all = array([1, 1, ..., 0, 0, 0, 0]) == Array([1, 1, ..., dtype=int32)
E
E             Full diff:
E             - Array([1, 1, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)
E             + array([1, 1, 1, 1, 1, 0, 0, 0, 0, 0]).all

tests/test_replay_buffer.py:121: AssertionError
______________ test_subtrajectory_replay_buffer_respects_episode_boundaries_on_termination _______________

strb_for_masking = <rl_blox.blox.replay_buffer.SubtrajectoryReplayBuffer object at 0x7caf84197560>

    def test_subtrajectory_replay_buffer_respects_episode_boundaries_on_termination(strb_for_masking):
        buf = SubtrajectoryReplayBuffer(buffer_size=10, horizon=3)
        for i in range(4): # all non-negative
            add(buf, obs=i, next_obs=i+1, term=False, trunc=False)
        add(buf, obs=4, next_obs=5, term=True, trunc=False) # still non-negative

        for i in range(-4, 0): # next episode, all negative
            add(buf, obs=i, next_obs=i-1, term=False, trunc=False)

        obs_subseqs, *_ = buf.sample_batch(
            batch_size=100,
            horizon=3,
            include_intermediate=True,
            rng=np.random.default_rng(0),
        )
        for obs_subseq in obs_subseqs:
>           assert (obs_subseq >= 0).all() or (obs_subseq <= 0).all()
E           assert (Array(False, dtype=bool) or Array(False, dtype=bool))
E            +  where Array(False, dtype=bool) = _all()
E            +    where _all = Array([ 4.,  5., -4.], dtype=float32) >= 0.all
E            +  and   Array(False, dtype=bool) = _all()
E            +    where _all = Array([ 4.,  5., -4.], dtype=float32) <= 0.all

tests/test_replay_buffer.py:145: AssertionError
```

## MRQ

The code for `SubtrajectoryReplayBuffer` was adapted from MRQ's `ReplayBuffer`. Similar tests confirm that this code shares the same behavior (see the tests [here](https://github.com/facebookresearch/MRQ/commit/1848ab902c1dfa9eceb6b9017b99afcca401ffa8)). Relevant test output:

```
______________________________ test_subtrajectory_replay_buffer_termination ______________________________

strb_for_masking = <MRQ.buffer.ReplayBuffer object at 0x7c57edee0770>

    def test_subtrajectory_replay_buffer_termination(strb_for_masking):
        strb_for_masking.add(
            state=torch.tensor([4.0]),
            action=12.0,
            reward=7,
            next_state=torch.tensor([5.0]),
            terminated=True,
            truncated=False,
        )
>       assert (strb_for_masking.mask == torch.tensor([1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])).all()
E       assert tensor(False)
E        +  where tensor(False) = <built-in method all of Tensor object at 0x7c56fa877f20>()
E        +    where <built-in method all of Tensor object at 0x7c56fa877f20> = tensor([1., 1..., 0., 0., 0.]) == tensor([1., 1..., 0., 0., 0.])
E
E             Full diff:
E             - tensor([1., 1., 0., 0., 0., 0., 0., 0., 0., 0.])
E             ?                                ------------
E             + tensor([1., 1., 1., 1., 1., 0., 0., 0., 0., 0.])
E             ?                 ++++++++++++.all

tests/test_replay_buffer.py:54: AssertionError
______________ test_subtrajectory_replay_buffer_respects_episode_boundaries_on_termination _______________

strb_for_masking = <MRQ.buffer.ReplayBuffer object at 0x7c57b4555220>

    def test_subtrajectory_replay_buffer_respects_episode_boundaries_on_termination(strb_for_masking):
        buffer = strb_for_masking # all non-negative
        buffer.add(
            state=torch.tensor([4.0]),
            action=12.0,
            reward=7,
            next_state=torch.tensor([5.0]),
            terminated=True,
            truncated=False,
        ) # still non-negative
        for i in range(-4, 0): # next episode, all negative
            buffer.add(
                state=torch.tensor([float(i)]),
                action=12.0,
                reward=7,
                next_state=torch.tensor([i-1.0]),
                terminated=False,
                truncated=False,
            )
        obs_subseqs, *_ = buffer.sample(
            horizon=3,
            include_intermediate=True,
        )
        for obs_subseq in obs_subseqs:
>           assert (obs_subseq >= 0).all() or (obs_subseq <= 0).all()
E           assert (tensor(False) or tensor(False))
E            +  where tensor(False) = <built-in method all of Tensor object at 0x7c56fa8dccd0>()
E            +    where <built-in method all of Tensor object at 0x7c56fa8dccd0> = tensor([[ 4.],\n        [ 0.],\n        [-4.]]) >= 0.all
E            +  and   tensor(False) = <built-in method all of Tensor object at 0x7c56fa8dcdc0>()
E            +    where <built-in method all of Tensor object at 0x7c56fa8dcdc0> = tensor([[ 4.],\n        [ 0.],\n        [-4.]]) <= 0.all

tests/test_replay_buffer.py:91: AssertionError
```

**Comment by adrian-schucht** [comment]
As a fix, I propose not setting the mask to 1 in the non-truncated (i.e., terminated) case (see second commit of #212). While I do not see a use for the remaining setting the mask to 0 in the truncated case, it does not seem to cause problems and removing it would risk breaking it in case it does have a use.
